aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -384,6 +374,153 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
+    TORCH_CHECK(self.layout() == Layout::Strided,
+                "clamp only supports strided layout, got: ", self.layout());
+
+    // We clamp max tensor to make sure that it is not smaller than min
+    // This is necessary to fix the divergent behaviour in AVX and base CPU kernels
+    // https://github.com/pytorch/pytorch/pull/32587#issuecomment-591389582
+    auto adjusted_max = at::clamp_with_tensors_min(max, min);
+
+    auto iter = TensorIterator();
+    iter.set_check_mem_overlap(true);
+    iter.add_output(result);
+    iter.add_input(self);
+    iter.add_input(min);
+    iter.add_input(adjusted_max);
+    iter.promote_common_dtype();
+    iter.build();
+    clamp_with_tensors_stub(iter.device_type(), iter);
+  } else if (max.defined()) {
+    at::clamp_with_tensors_max_out(result, self, max);
+  } else if (min.defined()) {
+    at::clamp_with_tensors_min_out(result, self, min);
+  } else {
+    AT_ERROR("At least one of 'min' or 'max' must not be None");
+  }
+  return result;
+}
+
+Tensor& clamp_with_tensors_max_(Tensor& self, const Tensor& max) {
+  return at::clamp_with_tensors_max_out(self, self, max);
+}
+
+Tensor& clamp_with_tensors_max_out(Tensor& result, const Tensor& self, const Tensor& max) {
+  TORCH_CHECK(!self.is_complex(), "clamp is not yet implemented for complex tensors.");
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "clamp_max only supports CPU or CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "clamp_max only supports strided layout, got: ", self.layout());
+  auto iter = TensorIterator();
+  iter.set_check_mem_overlap(true);
+  iter.add_output(result);
+  iter.add_input(self);
+  iter.add_input(max);
+  iter.promote_common_dtype();
+  iter.build();
+  clamp_max_with_tensor_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor& clamp_with_tensors_min_(Tensor& self, const Tensor& min) {
+  return at::clamp_with_tensors_min_out(self, self, min);
+}
+
+Tensor& clamp_with_tensors_min_out(Tensor& result, const Tensor& self, const Tensor& min) {
+  TORCH_CHECK(!self.is_complex(), "clamp is not yet implemented for complex tensors.");
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "clamp_min only supports CPU or CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "clamp_min only supports strided layout, got: ", self.layout());
+  auto iter = TensorIterator();
+  iter.set_check_mem_overlap(true);
+  iter.add_output(result);
+  iter.add_input(self);
+  iter.add_input(min);
+  iter.promote_common_dtype();
+  iter.build();
+  clamp_min_with_tensor_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor clamp_with_min_tensor(const Tensor& self, const Tensor& min, Scalar max) {
```

Comment:
You might be able to combine some of these function definitions by optionally taking arguments like max, which would then be given as a c10::optional that you could query for whether it has a value or not.

Reply:
Hmm. I remember trying to combine them a while ago, but if I make `max` optional then the signatures in `native_functions.yaml` would conflict for function definitions: `clamp_with_tensors(Tensor self, Tensor? min=None, Tensor? max=None)` and `clamp_with_tensors.min_tensor(Tensor self, Tensor min, Scalar? max=None)`. I.e. it couldn't distinguish between a call to `clamp_with_tensors` with `min` argument only and `clamp_with_tensors.min_tensor` without `max`, and hence it complained at compile time. 

Secondly, if I were to reuse a cpp function, for example, `clamp_with_tensors_min(const Tensor& self, const Tensor& min, optional<Scalar> max)` for `native_functions.yaml` signatures `clamp_with_tensors.min_tensor(Tensor self, Tensor min, Scalar max)` and `clamp_with_tensors_min(Tensor self, Tensor min)`, how would I attach a dispatch for `clamp_with_tensors_min` to call cpp function `clamp_with_tensors_min`? I think the build would also complain regarding signature mismatch. 

I suppose the only alternative option is to change the definition of `clamp_with_tensors_min(Tensor self, Tensor min)` to `clamp_with_tensors_min(Tensor self, Tensor min, Scalar? max=None)`, and get rid of the extra `clamp_with_tensors.min_tensor` native_functions signature and the `clamp_with_min_tensor` cpp function you highlighted. However, in this case a call to `torch.clamp(min_tensor, max_scalar)` will not get automatically dispatched using the native dynamic dispatch. I would thus have to add an extra type check in `functional.py` in order to dispatch to the correct aten function.

Neither option seems perfect, so I would appreciate your thoughts.

Reply:
I appreciate your thorough thinking, and confess I don't have a great idea for how to resolve the issue at this time, especially since our current clamp implementation has similar oddities. I think we can leave it for a follow-up PR.